File: src/routes/question/question.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common'
import { QuestionRepository } from './question.repo'
import { CreateQuestionDTO, UpdateQuestionDTO, QueryQuestionDTO, BulkCreateQuestionsDTO } from './question.dto'
import { QuestionWhereInput, QuestionOrderByInput, QuestionWithOptions } from './question.model'
import { JLPTLevelType, QuestionTypeType, DifficultyType } from 'src/shared/constants/enum.constant'
import { QuestionType } from 'src/shared/types/question.types'
import { S3Service } from 'src/shared/services/s3.service'

@Injectable()
export class QuestionService {
  constructor(
    private readonly questionRepository: QuestionRepository,
    private readonly s3Service: S3Service,
  ) {}

  async create(
    createDto: CreateQuestionDTO,
    files?: { image?: Express.Multer.File[]; audio?: Express.Multer.File[] },
  ): Promise<any> {
    const { mediaId, ...questionData } = createDto

    // Handle file uploads
    let uploadedMediaId = mediaId

    if (files?.image?.[0] || files?.audio?.[0]) {
      // Upload files to S3 and create media record
      let imageUrl: string | undefined
      let audioUrl: string | undefined
      let imageFile: Express.Multer.File | undefined
      let audioFile: Express.Multer.File | undefined

      if (files.image?.[0]) {
        const imageResult = await this.s3Service.uploadFileToS3(files.image[0], 'questions/images')
        imageUrl = imageResult.url
        imageFile = files.image[0]
      }

      if (files.audio?.[0]) {
        const audioResult = await this.s3Service.uploadFileToS3(files.audio[0], 'questions/audio')
        audioUrl = audioResult.url
        audioFile = files.audio[0]
      }

      // Create media record in database with both files info
      if (imageUrl || audioUrl) {
        // Primary URL is image if available, otherwise audio
        const primaryUrl = imageUrl || audioUrl!
        const primaryKind = imageUrl ? 'IMAGE' : 'AUDIO'

        // Build metadata to store info about both files
        const metadata: any = {}
        if (imageUrl && audioUrl) {
          // Both files uploaded
          metadata.hasMultiple = true
          metadata.image = {
            url: imageUrl,
            mimeType: imageFile?.mimetype,
            sizeByte: imageFile?.size,
          }
          metadata.audio = {
            url: audioUrl,
            mimeType: audioFile?.mimetype,
            sizeByte: audioFile?.size,
          }
        }

        const mediaData = {
          url: primaryUrl,
          kind: primaryKind,
          caption: null,
          mimeType: imageFile?.mimetype || audioFile?.mimetype,
          sizeByte: imageFile?.size || audioFile?.size,
          meta: Object.keys(metadata).length > 0 ? metadata : null,
        }

        const media = await this.questionRepository.createMedia(mediaData)
        uploadedMediaId = media.id
      }
    }

    if (uploadedMediaId) {
      const mediaExists = await this.questionRepository.checkMediaExists(uploadedMediaId)
      if (!mediaExists) {
        throw new BadRequestException(`Media with ID ${uploadedMediaId} does not exist`)
      }
    }

    const questionCreateData: any = {
      ...questionData,
    }

    if (uploadedMediaId) {
      questionCreateData.media = { connect: { id: uploadedMediaId } }
    }

    return this.questionRepository.create(questionCreateData)
  }

  async findAll(queryDto: QueryQuestionDTO): Promise<{
    data: any[]
    pagination: {
      total: number
      page: number
      limit: number
      totalPages: number
      hasNext: boolean
      hasPrev: boolean
    }
  }> {
    const {
      page = 1,
      limit = 10,
      type,
      level,
      difficulty,
      readingLength,
      keyword,
      hasMedia,
      sortBy = 'createdAt',
      sortOrder = 'desc',
    } = queryDto

    if (page < 1) {
      throw new BadRequestException('Page must be greater than 0')
    }

    if (limit < 1 || limit > 100) {
      throw new BadRequestException('Limit must be between 1 and 100')
    }

    const skip = (page - 1) * limit

    const where: QuestionWhereInput = {}

    if (type !== undefined) {
      where.type = type
    }

    if (level !== undefined) {
      where.level = level
    }

    if (difficulty !== undefined) {
      where.difficulty = difficulty
    }

    // Note: readingLength filter may need adjustment based on Prisma schema
    // if (readingLength !== undefined && readingLength !== null) {
    //   where.readingLength = readingLength
    // }

    if (keyword) {
      where.OR = [
        { stem: { contains: keyword, mode: 'insensitive' } },
        { passage: { contains: keyword, mode: 'insensitive' } },
        { explanation: { contains: keyword, mode: 'insensitive' } },
      ]
    }

    if (hasMedia !== undefined) {
      where.mediaId = hasMedia ? { not: null } : null
    }

    // Build order by clause
    const orderBy: QuestionOrderByInput = {}
    if (sortBy && sortOrder) {
      orderBy[sortBy] = sortOrder
    }

    try {
      const [questions, total] = await Promise.all([
        this.questionRepository.findManyWithStats({
          skip,
          take: limit,
          where,
          orderBy,
        }),
        this.questionRepository.count(where),
      ])

      // Transform questions to include stats
      const transformedQuestions = questions.map((question) => ({
        ...question,
        correctOptionsCount: question.option?.filter((opt) => opt.isCorrect).length || 0,
        hasMedia: !!question.mediaId,
      }))

      return {
        data: transformedQuestions,
        pagination: {
          total,
          page,
          limit,
          totalPages: Math.ceil(total / limit),
          hasNext: page * limit < total,
          hasPrev: page > 1,
        },
      }
    } catch (error) {
      throw new BadRequestException(
        `Failed to fetch questions: ${error instanceof Error ? error.message : 'Unknown error'}`,
      )
    }
  }

  async findOne(id: number): Promise<any> {
    const question = await this.questionRepository.findUnique({ id }, true)
    if (!question) {
      throw new NotFoundException(`Question with ID ${id} not found`)
    }
    return question
  }

  async update(
    id: number,
    updateDto: UpdateQuestionDTO,
    files?: { image?: Express.Multer.File[]; audio?: Express.Multer.File[] },
  ): Promise<any> {
    const exists = await this.questionRepository.checkExists(id)
    if (!exists) {
      throw new NotFoundException(`Question with ID ${id} not found`)
    }

    const isUsed = await this.questionRepository.isQuestionUsed(id)
    if (isUsed) {
      throw new BadRequestException(
        'Cannot update this question as it is already used in assessments or quizzes. Please create a new version instead using the clone endpoint.',
      )
    }

    const { mediaId, ...questionData } = updateDto

    let updatedMediaId = mediaId

    if (files?.image?.[0] || files?.audio?.[0]) {
      // Upload files to S3 and create media record
      let imageUrl: string | undefined
      let audioUrl: string | undefined
      let imageFile: Express.Multer.File | undefined
      let audioFile: Express.Multer.File | undefined

      if (files.image?.[0]) {
        const imageResult = await this.s3Service.uploadFileToS3(files.image[0], 'questions/images')
        imageUrl = imageResult.url
        imageFile = files.image[0]
      }

      if (files.audio?.[0]) {
        const audioResult = await this.s3Service.uploadFileToS3(files.audio[0], 'questions/audio')
        audioUrl = audioResult.url
        audioFile = files.audio[0]
      }

      // Create media record in database with both files info
      if (imageUrl || audioUrl) {
        // Primary URL is image if available, otherwise audio
        const primaryUrl = imageUrl || audioUrl!
        const primaryKind = imageUrl ? 'IMAGE' : 'AUDIO'

        // Build metadata to store info about both files
        const metadata: any = {}
        if (imageUrl && audioUrl) {
          // Both files uploaded
          metadata.hasMultiple = true
          metadata.image = {
            url: imageUrl,
            mimeType: imageFile?.mimetype,
            sizeByte: imageFile?.size,
          }
          metadata.audio = {
            url: audioUrl,
            mimeType: audioFile?.mimetype,
            sizeByte: audioFile?.size,
          }
        }

        const mediaData = {
          url: primaryUrl,
          kind: primaryKind,
          caption: null,
          mimeType: imageFile?.mimetype || audioFile?.mimetype,
          sizeByte: imageFile?.size || audioFile?.size,
          meta: Object.keys(metadata).length > 0 ? metadata : null,
        }

        const media = await this.questionRepository.createMedia(mediaData)
        updatedMediaId = media.id
      }
    }

    if (updatedMediaId) {
      const mediaExists = await this.questionRepository.checkMediaExists(updatedMediaId)
      if (!mediaExists) {
        throw new BadRequestException(`Media with ID ${updatedMediaId} does not exist`)
      }
    }

    const updateData: any = {
      ...questionData,
    }

    if (updatedMediaId) {
      updateData.media = { connect: { id: updatedMediaId } }
    }

    return this.questionRepository.update({ id }, updateData)
  }

  async remove(id: number): Promise<any> {
    const exists = await this.questionRepository.checkExists(id)
    if (!exists) {
      throw new NotFoundException(`Question with ID ${id} not found`)
    }

    // Check if question is used in any assessment or quiz
    const isUsed = await this.questionRepository.isQuestionUsed(id)
    if (isUsed) {
      throw new BadRequestException(
        'Cannot delete this question as it is already used in assessments or quizzes. Questions in use are protected from deletion.',
      )
    }

    return this.questionRepository.delete({ id })
  }

  async getStatistics() {
    return this.questionRepository.getStatistics()
  }

  async bulkCreate(bulkCreateDto: BulkCreateQuestionsDTO) {
    const { questions } = bulkCreateDto

    for (const question of questions) {
      if (question.mediaId) {
        const mediaExists = await this.questionRepository.checkMediaExists(question.mediaId)
        if (!mediaExists) {
          throw new BadRequestException(`Media with ID ${question.mediaId} does not exist`)
        }
      }

      const correctOptions = question.options.filter((opt) => opt.isCorrect)
      if (correctOptions.length === 0) {
        throw new BadRequestException('At least one option must be correct for each question')
      }
    }

    const questionsData = questions.map((question) => {
      const { options, mediaId, ...questionData } = question

      const questionCreateData: any = {
        ...questionData,
      }

      if (mediaId) {
        questionCreateData.media = { connect: { id: mediaId } }
      }

      return {
        questionData: questionCreateData,
        options: options.map((opt, index) => ({
          content: opt.content,
          isCorrect: opt.isCorrect,
          order: opt.order ?? index,
        })),
      }
    })

    return this.questionRepository.bulkCreate(questionsData)
  }

  async findByType(type: QuestionTypeType, queryDto: Omit<QueryQuestionDTO, 'type'>) {
    const queryWithType = { ...queryDto, type }
    return this.findAll(queryWithType as any)
  }

  async findByLevel(level: JLPTLevelType, queryDto: Omit<QueryQuestionDTO, 'level'>) {
    const queryWithLevel = { ...queryDto, level }
    return this.findAll(queryWithLevel)
  }

  async findByDifficulty(difficulty: DifficultyType, queryDto: Omit<QueryQuestionDTO, 'difficulty'>) {
    const queryWithDifficulty = { ...queryDto, difficulty }
    return this.findAll(queryWithDifficulty)
  }

  // Additional helper methods for consistency with test-section pattern
  async getQuestionsByFilters(filters: {
    type?: QuestionTypeType
    level?: JLPTLevelType
    difficulty?: DifficultyType
    page?: number
    limit?: number
    sortBy?: string
    sortOrder?: string
  }) {
    const result = await this.questionRepository.findByFilters(filters as any)

    const transformedData = result.data.map((question) => ({
      ...question,
      optionsCount: question._count?.option || 0,
      correctOptionsCount: question.option?.filter((opt: any) => opt.isCorrect).length || 0,
      hasMedia: !!question.mediaId,
    }))

    return {
      data: transformedData,
      pagination: {
        total: result.total,
        page: filters.page || 1,
        limit: filters.limit || 10,
        totalPages: Math.ceil(result.total / (filters.limit || 10)),
        hasNext: (filters.page || 1) * (filters.limit || 10) < result.total,
        hasPrev: (filters.page || 1) > 1,
      },
    }
  }

  // ============= VERSIONING METHODS =============

  /**
   * Clone a question to create a new version with optional modifications
   */
  async cloneQuestion(
    id: number,
    modifications?: {
      type?: QuestionTypeType
      level?: JLPTLevelType
      difficulty?: DifficultyType
      stem?: string
      passage?: string | null
      explanation?: string | null
      readingLength?: string | null
      mediaId?: number | null
      options?: Array<{
        content: string
        isCorrect: boolean
        order: number
        mediaId?: number | null
      }>
    },
  ): Promise<any> {
    const exists = await this.questionRepository.checkExists(id)
    if (!exists) {
      throw new NotFoundException(`Question with ID ${id} not found`)
    }

    // Validate mediaId if provided
    if (modifications?.mediaId) {
      const mediaExists = await this.questionRepository.checkMediaExists(modifications.mediaId)
      if (!mediaExists) {
        throw new BadRequestException(`Media with ID ${modifications.mediaId} does not exist`)
      }
    }

    // Validate options if provided
    if (modifications?.options) {
      const correctOptions = modifications.options.filter((opt) => opt.isCorrect)
      if (correctOptions.length === 0) {
        throw new BadRequestException('At least one option must be correct')
      }

      if (modifications.options.length < 2) {
        throw new BadRequestException('At least 2 options are required')
      }

      if (modifications.options.length > 6) {
        throw new BadRequestException('Maximum 6 options allowed')
      }
    }

    return this.questionRepository.cloneQuestion(id, modifications)
  }

  /**
   * Get all versions of a question by UUID
   */
  async getQuestionVersions(uuid: string): Promise<any[]> {
    const versions = await this.questionRepository.findVersionsByUuid(uuid)
    if (!versions || versions.length === 0) {
      throw new NotFoundException(`No questions found with UUID ${uuid}`)
    }

    return versions.map((question) => ({
      ...question,
      optionsCount: question.option?.length || 0,
      correctOptionsCount: question.option?.filter((opt: any) => opt.isCorrect).length || 0,
      hasMedia: !!question.mediaId,
    }))
  }

  /**
   * Get a specific version of a question
   */
  async getQuestionByVersion(uuid: string, version: number): Promise<any> {
    const question = await this.questionRepository.findByUuidAndVersion(uuid, version)
    if (!question) {
      throw new NotFoundException(`Question with UUID ${uuid} and version ${version} not found`)
    }

    return {
      ...question,
      optionsCount: question.option?.length || 0,
      correctOptionsCount: question.option?.filter((opt: any) => opt.isCorrect).length || 0,
      hasMedia: !!question.mediaId,
    }
  }

  /**
   * Check if a question can be edited or deleted
   */
  async checkQuestionUsage(id: number): Promise<{
    canEdit: boolean
    canDelete: boolean
    usageDetails: {
      inAssessments: boolean
      inQuizzes: boolean
      hasAnswers: boolean
    }
  }> {
    const exists = await this.questionRepository.checkExists(id)
    if (!exists) {
      throw new NotFoundException(`Question with ID ${id} not found`)
    }

    const isUsed = await this.questionRepository.isQuestionUsed(id)

    // For detailed breakdown, we would need additional queries
    // For now, returning a simple response
    return {
      canEdit: !isUsed,
      canDelete: !isUsed,
      usageDetails: {
        inAssessments: isUsed,
        inQuizzes: isUsed,
        hasAnswers: isUsed,
      },
    }
  }
}

File: assessment-service/src/main/java/com/torii/assessment/service/QuestionService.java
package com.torii.assessment.service;

import com.torii.assessment.dto.question.*;
import com.torii.assessment.entity.Option;
import com.torii.assessment.entity.Question;
import com.torii.assessment.repository.OptionRepository;
import com.torii.assessment.repository.QuestionRepository;
import jakarta.persistence.criteria.Predicate;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class QuestionService {

    private final QuestionRepository questionRepository;
    private final OptionRepository optionRepository;
    private final S3Service s3Service;

    @Transactional
    public QuestionResponseDTO createQuestion(CreateQuestionDTO dto, MultipartFile image, MultipartFile audio) {
        // Upload files to S3 if provided
        String mediaUrl = null;
        try {
            if (image != null && !image.isEmpty()) {
                mediaUrl = s3Service.uploadFile(image, "questions/images");
                log.info("Uploaded image: {}", mediaUrl);
            } else if (audio != null && !audio.isEmpty()) {
                mediaUrl = s3Service.uploadFile(audio, "questions/audio");
                log.info("Uploaded audio: {}", mediaUrl);
            }
        } catch (IOException e) {
            log.error("Error uploading file: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to upload file", e);
        }

        // Use uploaded URL or provided mediaId
        String finalMediaUrl = mediaUrl;

        // Validate options
        if (dto.getOptions() != null && !dto.getOptions().isEmpty()) {
            long correctCount = dto.getOptions().stream()
                .filter(opt -> Boolean.TRUE.equals(opt.getIsCorrect()))
                .count();
            if (correctCount < 1) {
                throw new IllegalArgumentException("At least one option must be correct");
            }
        }

        Question question = Question.builder()
            .type(dto.getType())
            .level(dto.getLevel())
            .difficulty(dto.getDifficulty() != null ? dto.getDifficulty() : Question.Difficulty.MEDIUM)
            .stem(dto.getStem())
            .passage(dto.getPassage())
            .mediaId(dto.getMediaId())
            .mediaUrl(finalMediaUrl)
            .explanation(dto.getExplanation())
            .readingLength(dto.getReadingLength())
            .build();

        Question saved = questionRepository.save(question);
        log.info("Created question: {}", saved.getId());

        // Create options
        if (dto.getOptions() != null && !dto.getOptions().isEmpty()) {
            List<Option> options = new ArrayList<>();
            for (int i = 0; i < dto.getOptions().size(); i++) {
                CreateQuestionDTO.CreateOptionDTO optDto = dto.getOptions().get(i);
                Option option = Option.builder()
                    .questionId(saved.getId())
                    .content(optDto.getContent())
                    .isCorrect(optDto.getIsCorrect() != null ? optDto.getIsCorrect() : false)
                    .order(optDto.getOrder() != null ? optDto.getOrder() : i)
                    .mediaId(optDto.getMediaId())
                    .build();
                options.add(option);
            }
            optionRepository.saveAll(options);
        }

        return getQuestionById(saved.getId());
    }

    public Map<String, Object> getAllQuestions(QueryQuestionDTO queryDto) {
        int page = queryDto.getPage() != null ? queryDto.getPage() : 1;
        int limit = queryDto.getLimit() != null ? queryDto.getLimit() : 10;

        if (page < 1) {
            throw new IllegalArgumentException("Page must be greater than 0");
        }
        if (limit < 1 || limit > 100) {
            throw new IllegalArgumentException("Limit must be between 1 and 100");
        }

        Sort sort = Sort.by(
            "desc".equalsIgnoreCase(queryDto.getSortOrder()) ? Sort.Direction.DESC : Sort.Direction.ASC,
            queryDto.getSortBy() != null ? queryDto.getSortBy() : "createdAt"
        );
        Pageable pageable = PageRequest.of(page - 1, limit, sort);

        Specification<Question> spec = buildSpecification(queryDto);
        Page<Question> questionPage = questionRepository.findAll(spec, pageable);

        List<QuestionResponseDTO> data = questionPage.getContent().stream()
            .map(this::mapToResponseDTO)
            .collect(Collectors.toList());

        Map<String, Object> result = new HashMap<>();
        result.put("data", data);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("total", questionPage.getTotalElements());
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("totalPages", questionPage.getTotalPages());
        pagination.put("hasNext", questionPage.hasNext());
        pagination.put("hasPrev", questionPage.hasPrevious());
        result.put("pagination", pagination);

        return result;
    }

    public QuestionResponseDTO getQuestionById(Long id) {
        Question question = questionRepository.findByIdWithOptions(id)
            .orElseThrow(() -> new RuntimeException("Question not found: " + id));
        return mapToResponseDTO(question);
    }

    @Transactional
    public QuestionResponseDTO updateQuestion(Long id, UpdateQuestionDTO dto, MultipartFile image, MultipartFile audio) {
        Question question = questionRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Question not found: " + id));

        // Upload new files if provided
        String newMediaUrl = null;
        try {
            if (image != null && !image.isEmpty()) {
                // Delete old media if exists
                if (question.getMediaUrl() != null && !question.getMediaUrl().isEmpty()) {
                    try {
                        s3Service.deleteFile(question.getMediaUrl());
                    } catch (Exception e) {
                        log.warn("Failed to delete old media: {}", e.getMessage());
                    }
                }
                newMediaUrl = s3Service.uploadFile(image, "questions/images");
                log.info("Uploaded new image: {}", newMediaUrl);
            } else if (audio != null && !audio.isEmpty()) {
                // Delete old media if exists
                if (question.getMediaUrl() != null && !question.getMediaUrl().isEmpty()) {
                    try {
                        s3Service.deleteFile(question.getMediaUrl());
                    } catch (Exception e) {
                        log.warn("Failed to delete old media: {}", e.getMessage());
                    }
                }
                newMediaUrl = s3Service.uploadFile(audio, "questions/audio");
                log.info("Uploaded new audio: {}", newMediaUrl);
            }
        } catch (IOException e) {
            log.error("Error uploading file: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to upload file", e);
        }

        // Update fields if provided
        if (dto.getType() != null) question.setType(dto.getType());
        if (dto.getLevel() != null) question.setLevel(dto.getLevel());
        if (dto.getDifficulty() != null) question.setDifficulty(dto.getDifficulty());
        if (dto.getStem() != null) question.setStem(dto.getStem());
        if (dto.getPassage() != null) question.setPassage(dto.getPassage());
        if (newMediaUrl != null) {
            question.setMediaUrl(newMediaUrl);
        }
        if (dto.getMediaId() != null) {
            question.setMediaId(dto.getMediaId());
        }
        if (dto.getExplanation() != null) question.setExplanation(dto.getExplanation());
        if (dto.getReadingLength() != null) question.setReadingLength(dto.getReadingLength());

        questionRepository.save(question);
        log.info("Updated question: {}", id);

        // Update options if provided
        if (dto.getOptions() != null && !dto.getOptions().isEmpty()) {
            // Validate at least one correct
            long correctCount = dto.getOptions().stream()
                .filter(opt -> Boolean.TRUE.equals(opt.getIsCorrect()))
                .count();
            if (correctCount < 1) {
                throw new IllegalArgumentException("At least one option must be correct");
            }

            // Delete existing options and create new ones
            optionRepository.deleteAllByQuestionId(id);

            List<Option> options = new ArrayList<>();
            for (int i = 0; i < dto.getOptions().size(); i++) {
                UpdateQuestionDTO.UpdateOptionDTO optDto = dto.getOptions().get(i);
                Option option = Option.builder()
                    .questionId(id)
                    .content(optDto.getContent())
                    .isCorrect(optDto.getIsCorrect() != null ? optDto.getIsCorrect() : false)
                    .order(optDto.getOrder() != null ? optDto.getOrder() : i)
                    .mediaId(optDto.getMediaId())
                    .build();
                options.add(option);
            }
            optionRepository.saveAll(options);
        }

        return getQuestionById(id);
    }

    @Transactional
    public void deleteQuestion(Long id) {
        if (!questionRepository.existsById(id)) {
            throw new RuntimeException("Question not found: " + id);
        }
        questionRepository.deleteById(id);
        log.info("Deleted question: {}", id);
    }

    public QuestionStatsDTO getStatistics() {
        long total = questionRepository.count();

        Map<String, Long> byType = new HashMap<>();
        for (Question.QuestionType type : Question.QuestionType.values()) {
            byType.put(type.name(), questionRepository.countByType(type));
        }

        Map<String, Long> byLevel = new HashMap<>();
        for (Question.JLPTLevel level : Question.JLPTLevel.values()) {
            byLevel.put(level.name(), questionRepository.countByLevel(level));
        }

        Map<String, Long> byDifficulty = new HashMap<>();
        for (Question.Difficulty difficulty : Question.Difficulty.values()) {
            byDifficulty.put(difficulty.name(), questionRepository.countByDifficulty(difficulty));
        }

        return QuestionStatsDTO.builder()
            .totalQuestions(total)
            .byType(byType)
            .byLevel(byLevel)
            .byDifficulty(byDifficulty)
            .withMedia(questionRepository.countByMediaIdIsNotNull())
            .withoutMedia(questionRepository.countByMediaIdIsNull())
            .build();
    }

    @Transactional
    public List<QuestionResponseDTO> bulkCreateQuestions(BulkCreateQuestionsDTO dto) {
        List<QuestionResponseDTO> results = new ArrayList<>();
        for (CreateQuestionDTO questionDto : dto.getQuestions()) {
            results.add(createQuestion(questionDto, null, null));
        }
        return results;
    }

    public Map<String, Object> findByType(Question.QuestionType type, QueryQuestionDTO queryDto) {
        queryDto.setType(type);
        return getAllQuestions(queryDto);
    }

    public Map<String, Object> findByLevel(Question.JLPTLevel level, QueryQuestionDTO queryDto) {
        queryDto.setLevel(level);
        return getAllQuestions(queryDto);
    }

    public Map<String, Object> findByDifficulty(Question.Difficulty difficulty, QueryQuestionDTO queryDto) {
        queryDto.setDifficulty(difficulty);
        return getAllQuestions(queryDto);
    }

    // Versioning methods
    public List<QuestionResponseDTO> getQuestionVersions(String uuid) {
        List<Question> versions = questionRepository.findByUuidOrderByVersionDesc(uuid);
        if (versions.isEmpty()) {
            throw new RuntimeException("No questions found with UUID: " + uuid);
        }
        return versions.stream()
            .map(this::mapToResponseDTO)
            .collect(Collectors.toList());
    }

    public QuestionResponseDTO getQuestionByVersion(String uuid, Integer version) {
        Question question = questionRepository.findByUuidAndVersion(uuid, version)
            .orElseThrow(() -> new RuntimeException("Question not found with UUID " + uuid + " and version " + version));
        return mapToResponseDTO(question);
    }

    @Transactional
    public QuestionResponseDTO cloneQuestion(Long id, UpdateQuestionDTO modifications) {
        Question original = questionRepository.findByIdWithOptions(id)
            .orElseThrow(() -> new RuntimeException("Question not found: " + id));

        // Get latest version
        Integer latestVersion = questionRepository.findByUuidOrderByVersionDesc(original.getUuid())
            .stream()
            .findFirst()
            .map(Question::getVersion)
            .orElse(0);

        Question clone = Question.builder()
            .uuid(original.getUuid())
            .version(latestVersion + 1)
            .type(modifications.getType() != null ? modifications.getType() : original.getType())
            .level(modifications.getLevel() != null ? modifications.getLevel() : original.getLevel())
            .difficulty(modifications.getDifficulty() != null ? modifications.getDifficulty() : original.getDifficulty())
            .stem(modifications.getStem() != null ? modifications.getStem() : original.getStem())
            .passage(modifications.getPassage() != null ? modifications.getPassage() : original.getPassage())
            .mediaId(modifications.getMediaId() != null ? modifications.getMediaId() : original.getMediaId())
            .mediaUrl(original.getMediaUrl())
            .explanation(modifications.getExplanation() != null ? modifications.getExplanation() : original.getExplanation())
            .readingLength(modifications.getReadingLength() != null ? modifications.getReadingLength() : original.getReadingLength())
            .build();

        Question saved = questionRepository.save(clone);

        // Clone or create options
        List<Option> originalOptions = optionRepository.findByQuestionIdOrderByOrderAsc(id);
        if (modifications.getOptions() != null && !modifications.getOptions().isEmpty()) {
            for (int i = 0; i < modifications.getOptions().size(); i++) {
                UpdateQuestionDTO.UpdateOptionDTO optDto = modifications.getOptions().get(i);
                Option option = Option.builder()
                    .questionId(saved.getId())
                    .content(optDto.getContent())
                    .isCorrect(optDto.getIsCorrect() != null ? optDto.getIsCorrect() : false)
                    .order(optDto.getOrder() != null ? optDto.getOrder() : i)
                    .mediaId(optDto.getMediaId())
                    .build();
                optionRepository.save(option);
            }
        } else {
            for (Option opt : originalOptions) {
                Option clonedOpt = Option.builder()
                    .questionId(saved.getId())
                    .content(opt.getContent())
                    .isCorrect(opt.getIsCorrect())
                    .order(opt.getOrder())
                    .mediaId(opt.getMediaId())
                    .build();
                optionRepository.save(clonedOpt);
            }
        }

        log.info("Cloned question {} to new version {}", id, saved.getId());
        return getQuestionById(saved.getId());
    }

    public Map<String, Object> checkQuestionUsage(Long id) {
        if (!questionRepository.existsById(id)) {
            throw new RuntimeException("Question not found: " + id);
        }

        // TODO: Check actual usage in assessments and quizzes
        boolean isUsed = false;

        Map<String, Object> result = new HashMap<>();
        result.put("canEdit", !isUsed);
        result.put("canDelete", !isUsed);

        Map<String, Boolean> usageDetails = new HashMap<>();
        usageDetails.put("inAssessments", isUsed);
        usageDetails.put("inQuizzes", isUsed);
        usageDetails.put("hasAnswers", isUsed);
        result.put("usageDetails", usageDetails);

        return result;
    }

    // Helper methods
    private Specification<Question> buildSpecification(QueryQuestionDTO queryDto) {
        return (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();

            if (queryDto.getType() != null) {
                predicates.add(cb.equal(root.get("type"), queryDto.getType()));
            }
            if (queryDto.getLevel() != null) {
                predicates.add(cb.equal(root.get("level"), queryDto.getLevel()));
            }
            if (queryDto.getDifficulty() != null) {
                predicates.add(cb.equal(root.get("difficulty"), queryDto.getDifficulty()));
            }
            if (queryDto.getReadingLength() != null) {
                predicates.add(cb.equal(root.get("readingLength"), queryDto.getReadingLength()));
            }
            if (queryDto.getKeyword() != null && !queryDto.getKeyword().isEmpty()) {
                String keyword = "%" + queryDto.getKeyword().toLowerCase() + "%";
                Predicate stemPred = cb.like(cb.lower(root.get("stem")), keyword);
                Predicate passagePred = cb.like(cb.lower(root.get("passage")), keyword);
                Predicate expPred = cb.like(cb.lower(root.get("explanation")), keyword);
                predicates.add(cb.or(stemPred, passagePred, expPred));
            }
            if (queryDto.getHasMedia() != null) {
                if (queryDto.getHasMedia()) {
                    predicates.add(cb.or(
                        cb.isNotNull(root.get("mediaId")),
                        cb.isNotNull(root.get("mediaUrl"))
                    ));
                } else {
                    predicates.add(cb.and(
                        cb.isNull(root.get("mediaId")),
                        cb.isNull(root.get("mediaUrl"))
                    ));
                }
            }

            return cb.and(predicates.toArray(new Predicate[0]));
        };
    }

    private QuestionResponseDTO mapToResponseDTO(Question question) {
        List<Option> options = optionRepository.findByQuestionIdOrderByOrderAsc(question.getId());

        List<QuestionResponseDTO.OptionDTO> optionDTOs = options.stream()
            .map(opt -> QuestionResponseDTO.OptionDTO.builder()
                .id(opt.getId())
                .content(opt.getContent())
                .isCorrect(opt.getIsCorrect())
                .order(opt.getOrder())
                .mediaId(opt.getMediaId())
                .mediaUrl(opt.getMediaUrl())
                .build())
            .collect(Collectors.toList());

        long correctCount = options.stream()
            .filter(opt -> Boolean.TRUE.equals(opt.getIsCorrect()))
            .count();

        return QuestionResponseDTO.builder()
            .id(question.getId())
            .uuid(question.getUuid())
            .version(question.getVersion())
            .type(question.getType())
            .level(question.getLevel())
            .difficulty(question.getDifficulty())
            .stem(question.getStem())
            .passage(question.getPassage())
            .mediaId(question.getMediaId())
            .mediaUrl(question.getMediaUrl())
            .explanation(question.getExplanation())
            .readingLength(question.getReadingLength())
            .createdAt(question.getCreatedAt())
            .updatedAt(question.getUpdatedAt())
            .options(optionDTOs)
            .optionsCount(options.size())
            .correctOptionsCount((int) correctCount)
            .hasMedia(question.getMediaId() != null || question.getMediaUrl() != null)
            .build();
    }
}

